core/arch/arm/pta/device.c

```diff
@@ -0,0 +1,103 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2019, Linaro Limited
+ */
+
+/*
+ * This pseudo TA is used by normal world OS TEE driver to fetch pseudo TA's
+ * UUIDs which can act as TEE bus devices.
+ */
+
+#include <kernel/pseudo_ta.h>
+#include <kernel/tee_ta_manager.h>
+#include <tee/uuid.h>
+#include <pta_device.h>
+#include <string.h>
+#include <user_ta_header.h>
+
+#define PTA_NAME "device.pta"
+
+static TEE_Result get_num_devices(uint32_t types,
+				  TEE_Param params[TEE_NUM_PARAMS])
+{
+	const struct pseudo_ta_head *ta;
+	uint32_t num_devices = 0;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		EMSG("bad parameters types: 0x%" PRIx32, types);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	SCATTERED_ARRAY_FOREACH(ta, pseudo_tas, struct pseudo_ta_head) {
+		if (ta->flags & TA_FLAG_DEVICE_ENUM)
+			num_devices++;
+	}
+
+	params[0].value.a = num_devices;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result get_devices(uint32_t types,
+			      TEE_Param params[TEE_NUM_PARAMS])
+{
+	const struct pseudo_ta_head *ta;
+	TEE_UUID *device_uuid = NULL;
+	uint8_t uuid_octet[sizeof(TEE_UUID)];
+	size_t ip_size;
+	TEE_Result res = TEE_SUCCESS;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		EMSG("bad parameters types: 0x%" PRIx32, types);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (!params[0].memref.buffer) {
+		EMSG("bad parameters, memory buffer NULL\n");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	device_uuid = (TEE_UUID *)params[0].memref.buffer;
+	ip_size = params[0].memref.size;
+
+	SCATTERED_ARRAY_FOREACH(ta, pseudo_tas, struct pseudo_ta_head) {
+		if (ta->flags & TA_FLAG_DEVICE_ENUM) {
+			if (ip_size < sizeof(TEE_UUID)) {
+				res = TEE_ERROR_SHORT_BUFFER;
```

Comment:
Return code `TEE_ERROR_SHORT_BUFFER` expects that the required buffer size to returned to the client through the memref field `size`.
Here, you should set something like `params[0].memref.size = get_num_devices() * sizeof(TEE_UUID);`

Reply:
But why would client need this if there exist another invoke command: `PTA_CMD_GET_NUM_DEVICES` to get shm size?

Reply:
@b49020 please read GlobalPlatform Client API v1.0 section 3.2.5 *Memory References*, section *Variable Length Return Buffers*; as well as GlobalPlatform Internal Core API v1.1 section 3.4 *[outbuf]*, and drop `PTA_CMD_GET_NUM_DEVICES`.

Keeping two commands for this kind of operation is more complex, less efficient, and potentially wrong (racy).

Reply:
Thanks @jforissier for the references. So as per specification `memref=NULL or size=0` is an expected param, will update `copy_in_param()` (in `core/arch/arm/kernel/pseudo_ta.c`) accordingly. And also will drop `PTA_CMD_GET_NUM_DEVICES`.

Actually my initial expectation was that implementation in `pseudo_ta.c` is as per spec.